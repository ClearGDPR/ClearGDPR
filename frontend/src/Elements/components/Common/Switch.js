import React from 'react';
import PropTypes from 'prop-types';
import styles from '../../theme/Switch.scss';

class Switch extends React.PureComponent {
  constructor(props) {
    super(props);

    this.state = {
      value: props.value
    };
  }

  componentDidUpdate() {
    this.setState({ value: this.props.value });
  }

  handleClick() {
    if (this.props.disabled) {
      return;
    }
    this.setState(prevState => ({
      value: !prevState.value
    }));

    if (this.props.onChange) {
      this.props.onChange(!this.state.value);
    }
  }

  render() {
    const classes = [styles.switch, this.state.value ? styles.switchOn : ''].join(' ');

    return (
      <div className={classes} onClick={this.handleClick.bind(this)}>
        <div className={styles.switchToggle} />
      </div>
    );
  }
}

Switch.propTypes = {
  onChange: PropTypes.func.isRequired,
  value: PropTypes.bool,
  disabled: PropTypes.bool
};

export default Switch;
