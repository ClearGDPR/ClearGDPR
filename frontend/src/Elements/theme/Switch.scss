@import 'variables';

.switch {
  border: 1px solid #ccc;
  width: 50px;
  height: 26px;
  border-radius: 13px;
  cursor: pointer;
  display: inline-block;

  &On {
    background: $green;

    .switchToggle {
      left: 23px;
    }
  }

  &Disabled {
    cursor: not-allowed;
  }
}

.switchToggle {
  border: 1px solid #999;
  box-shadow: 1px 1px 1px #ccc;
  width: 25px;
  height: 24px;
  left: 0;
  border-radius: 12px;
  background: white;
  position: relative;
  transition: left 0.2s ease-in-out;
}
