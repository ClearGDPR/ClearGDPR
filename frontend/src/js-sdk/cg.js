import 'es6-promises';
import Config from './config';
import Modules from './modules';
import { SDKError } from './common/exceptions';
import { mergeOptions } from './utils/helpers';

/**
 * Creates a new CG ClearGDPR API client
 * @constructor
 * @classdesc The CG SDK class provides methods to manage user's data
 * @param {Object}  options
 * @param {String}  options.apiKey Key used to sign up to the API
 * @param {String}  options.apiUrl URL to the API server without trailing slash
 * @param {String}  options.apiVersion Version of the API, like `v1`
 * @param {String}  options.apiTimeout Timeout used when a Request is made to the API
 * @param {Boolean} options.debug Flag to enable or disable complete error stack trace
 * @see {@link https://docs.cleargdpr.com/api}
 */
export class CG {
  constructor({ apiKey, apiUrl }) {
    // Enforce accessToken option to be sent.
    if (!apiKey) {
      throw new SDKError('API Key is required.');
    }

    if (!apiUrl) {
      throw new SDKError('API URL is required.');
    }

    this._config = mergeOptions(Config, { apiUrl });
    this.setAPIKey(apiKey);
    this.registerModules();
  }

  registerModules() {
    Object.keys(Modules).map(key => {
      const endpoint = new Modules[key](this);
      Object.assign(this, { [key]: endpoint });
    });
  }

  getAccessToken() {
    return this._config.auth;
  }

  setAccessToken(accessToken) {
    if (accessToken) {
      // TODO: Should validate quicker if its a valid JWT
      // TODO: How to verify JWT with our Server + Consumer
      this._setConfigField('auth', `Bearer ${accessToken}`);
    }
  }

  getAPIKey() {
    return this._config.auth;
  }

  setAPIKey(apiKey) {
    if (apiKey) {
      this._setConfigField('apiKey', apiKey);
    }
  }

  getBaseURL() {
    return this._config.apiVersion
      ? `${this._config.apiUrl}/${this._config.apiVersion}`
      : this._config.apiUrl;
  }

  _setConfigField(key, value) {
    this._config[key] = value;
  }
}
