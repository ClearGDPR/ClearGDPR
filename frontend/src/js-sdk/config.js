const Config = {
  auth: null,
  apiKey: process.env.API_KEY,
  apiUrl: process.env.API_URL,
  apiVersion: process.env.API_VERSION,
  apiTimeout: 30,
  debug: false
};

export default Config;
