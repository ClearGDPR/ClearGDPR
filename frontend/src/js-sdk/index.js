/*
 * cg-js-sdk
 * https://www.cleargdpr.com/
 * 
 */
export { CG } from './cg';
