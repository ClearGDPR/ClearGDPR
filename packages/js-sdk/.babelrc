{
  "presets": [
    [
      "env",
      {
        "targets": {
          "browsers": ["> 1%", "last 2 versions"]
        }
      }
    ]
  ],
  "plugins": [
    [
      "module-resolver",
      {
        "extensions": [".js", ".jsx", ".es", ".es6", ".mjs"],
        "root": ["./src"],
        "alias": {}
      }
    ]
  ]
}
