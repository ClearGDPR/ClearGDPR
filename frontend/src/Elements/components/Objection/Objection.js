import React from 'react';
import PropTypes from 'prop-types';
import Switch from '../Common/Switch';

class Objection extends React.PureComponent {
  state = {
    allowDataProcessing: true,
    busy: true
  };
  async componentDidMount() {
    const { objection } = await this.props.cg.Subject.getObjectionStatus();
    this.setState({ busy: false, allowDataProcessing: !objection });
  }
  toggleDataProcessingAgreement = async () => {
    const { busy, allowDataProcessing } = this.state;
    if (busy) {
      return;
    }

    this.setState({ allowDataProcessing: !allowDataProcessing, busy: true });
    await this.props.cg.Subject.updateObjection(allowDataProcessing);
    this.setState({ busy: false });
  };
  render() {
    const { label = 'Object data processing' } = this.props.options;
    const { allowDataProcessing, busy } = this.state;

    return (
      <div style={{ alignItems: 'center', display: 'flex' }}>
        <label>{label}</label>&nbsp;
        <Switch
          onChange={this.toggleDataProcessingAgreement}
          value={allowDataProcessing}
          disabled={busy}
        />
      </div>
    );
  }
}

Objection.propTypes = {
  options: PropTypes.object,
  cg: PropTypes.object
};

export default Objection;
