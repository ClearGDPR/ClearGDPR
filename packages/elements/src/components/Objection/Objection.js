import React from 'react';
import PropTypes from 'prop-types';
import Switch from 'components/Common/Switch';
import Subject from 'contexts/Subject';

class Objection extends React.PureComponent {
  state = {
    allowDataProcessing: true,
    busy: false
  };
  componentDidMount() {
    this.props.subject.fetchObjectionStatus();
  }
  toggleDataProcessingAgreement = async () => {
    const { busy } = this.state;
    const { objection } = this.props.subject;
    if (busy) {
      return;
    }

    this.setState({ busy: true });
    await this.props.subject.updateObjection(!objection);
    this.setState({ busy: false });
  };
  render() {
    const { label = 'Object data processing' } = this.props.options;
    const { busy } = this.state;
    const { objection } = this.props.subject;

    return (
      <div style={{ alignItems: 'center', display: 'flex' }}>
        <label>{label}</label>
        &nbsp;
        <Switch
          onChange={this.toggleDataProcessingAgreement}
          value={!objection}
          disabled={busy || objection === null}
        />
      </div>
    );
  }
}

Objection.propTypes = {
  options: PropTypes.object,
  subject: PropTypes.instanceOf(Subject)
};

export default Objection;
