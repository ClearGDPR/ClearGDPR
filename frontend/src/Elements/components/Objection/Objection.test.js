import React from 'react';
import { mount } from 'enzyme';
import Objection from './Objection';
import { CG } from '../../../js-sdk';
import Switch from '../Common/Switch';

const cg = new CG({
  apiKey: 'test',
  apiUrl: 'test'
});
cg.Subject.getObjectionStatus = async () => ({ objection: true });
cg.Subject.updateObjection = async () => {};

const setup = () => mount(<Objection {...{ options: {}, cg }} />);

let spy;

describe('Objection', () => {
  beforeEach(() => {
    spy = null;
  });
  it('should render correctly', () => {
    const component = setup();
    expect(component).toMatchSnapshot();
  });

  it('should call `getObjectionStatus` method to retrieve current status', async () => {
    spy = jest.spyOn(cg.Subject, 'getObjectionStatus');
    setup();
    expect(spy).toHaveBeenCalled();
  });

  it('aligns state with received objection status', async () => {
    const component = await setup();
    const { allowDataProcessing, busy } = component.state();
    expect(allowDataProcessing).toBe(false);
    expect(busy).toBe(false);
  });

  it('should call `updateObjection` when subject toggles switch', async () => {
    spy = jest.spyOn(cg.Subject, 'updateObjection');
    const component = await setup();
    component.find(Switch).simulate('click');
    expect(component.state().allowDataProcessing).toBe(true);
    expect(spy).toHaveBeenCalledWith(false);
  });

  it('debounce switch toggles', async () => {
    spy = jest.spyOn(cg.Subject, 'updateObjection');
    const component = await setup();
    const switchComponent = component.find(Switch);
    switchComponent.simulate('click');
    switchComponent.simulate('click');
    expect(spy).toHaveBeenCalledTimes(1);
  });

  afterEach(() => {
    if (spy) {
      spy.mockClear();
    }
  });
});
