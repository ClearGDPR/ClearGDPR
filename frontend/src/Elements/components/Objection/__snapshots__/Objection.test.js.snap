// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Objection should render correctly 1`] = `Promise {}`;
